src/codegen/compile.ml

```diff
@@ -2350,6 +2353,59 @@ module MakeCompact (Num : BigNumType) : BigNumType = struct
           get_res
       end)
 
+
+  let compile_lsh env =
+    let set_amount, get_amount = new_local env "amount" in
+    let set_n, get_n = new_local env "n" in
+    set_amount ^^
+    set_n ^^ get_n ^^
+    BitTagged.if_tagged_scalar env [I32Type]
+      ( (* non-ptr, shift amount <= 42: signed i32 -> i64; shift left, remember, trunc to i32, extend, compare with recalled, if same, done! *)
+        get_n ^^
+        G.i (Convert (Wasm.Values.I64 I64Op.ExtendSI32)) ^^
+        get_amount ^^
+        G.i (Convert (Wasm.Values.I64 I64Op.ExtendUI32)) ^^
+        G.i (Binary (Wasm.Values.I64 I64Op.Shl)) ^^
+        let set_remember, get_remember = new_local64 env "remember" in
+        set_remember ^^ get_remember ^^
+        G.i (Convert (Wasm.Values.I32 I32Op.WrapI64)) ^^
+        let set_res, get_res = new_local env "res" in
+        set_res ^^ get_res ^^
+        G.i (Convert (Wasm.Values.I64 I64Op.ExtendSI32)) ^^ (* exclude sign flip *)
+        get_remember ^^
+        G.i (Compare (Wasm.Values.I64 I64Op.Eq)) ^^
```

Comment:
One could structure it in a way that we still reuse the 64-bit result (by going to heap from there) even if it fails to truncate to 32 bits.